### ISSUE TITLE
Accessing files of a custom drive from a kernel

### ISSUE THREAD

**Issue by jtpio** [jupyterlab/jupyterlab#15289]
### Problem

As noticed and discussed in https://github.com/jupyterlab/jupyterlab/pull/14519#issuecomment-1774055091, JupyterLab 4 does not store the drive name as a prefix on the session path anymore.

In JupyterLab 3, this drive name prefix used to be passed along when starting a new session with `/api/session`. That way the server could technically use this information and pass it to the kernel to allow accessing files from the kernel in a seamless way.

Also opening a notebook from a custom drive and executing the following cell would not list the content from the custom drive, which can be quite confusing to most users:

```py
import os
os.listdir()
```

### Proposed Solution

Drives are convenient ways to add more sources of contents to JupyterLab via frontend plugins. They also work well in other lab-based applications such as JupyterLite: https://github.com/jupyterlite/jupyterlite

Find a way to integrate the drive feature of JupyterLab so it plays well with other components of the Jupyter stacks.

### Additional context

Related discussions and issues:

- https://github.com/jupyterlab/jupyterlab/pull/14519#issuecomment-1774055091
- https://github.com/jupyterlite/jupyterlite/pull/1227
- https://github.com/jupyterlab/jupyterlab/issues/15143

**Comment by echarles** [comment]
> Also opening a notebook from a custom drive and executing the following cell would not list the content from the custom drive, which can be quite confusing to most users:

Let's imagine a `S3Drive` is defined on the frontend and its content is listed by the server content API. This leads me already to a first question: the relationship between the frontend drives and the content API.

The user double clicks on a listed notebook, and JupyterLab displays it and of course starts a python kernel. In you example, the òs.listdir()` python command run by the user in a S3 notebook cell will return the files and folders of the kernel, not the S3 files and folders... Second question: the relationship between the frontend drives and the kernels "drives" (mount point?).

What command would be supposed to list S3? In terms of user-experience, this is very confusing. I think we need to be clear on the boundaries between content API and kernel.

Being clear does not mean that we can not find a way to leverage the Frontend Drives in the Content API (enriching the API with an additional paramater?) and the Kernel (with mount points per drive??, the responsibility to mount them relying on the server/kernellauncher ?).

**Comment by davidbrochart** [comment]
> Let's imagine a `S3Drive` is defined on the frontend and its content is listed by the server content API. This leads me already to a first question: the relationship between the frontend drives and the content API.

A frontend drive can use the contents API or not. For instance, it could directly call an S3 API without passing through the Jupyter contents API, or it could call the contents API which in turn calls the S3 API.

> Second question: the relationship between the frontend drives and the kernels "drives" (mount point?).

There is no relationship in general, although in the common case where the frontend drive calls the Jupyter contents API, and the Jupyter server runs on the same machine as the kernels, the kernels can see the same file system.

I opened an [issue in jupyter-server](https://github.com/jupyter-server/jupyter_server/issues/1115) about exposing the contents API to kernels. A kernel could be a server client, and call the contents API. That would be a language-agnostic solution, but the kernel needs to be authenticated/authorised to call this API.

**Comment by davidbrochart** [comment]
I think that if we want to support the case where a kernel accesses a JupyterLab frontend drive that may not call the Jupyter contents API, the solution could be in the form of a Jupyter widget that would have access to the JupyterLab frontend drive. But it would be JupyterLab specific.

**Comment by krassowski** [comment]
For an exploration of the opposite problem - accessing/finding out who (server? client? remote drive?) owns files from the frontend space - see  https://github.com/jupyter-server/jupyter_server/issues/1280.

I feel that it is closely related because both could be solved with a new API/bridge between kernel and the filesystem owners (be it server's contents or the IDrive), with a centralised registry of what filesystems are available (and to whom).

This registry + API could allow to expose the filesystems to kernels, and to answer the question "who a file with this path" (even if multiple answers exist).

At the end of the day the question is who owns such a registry (server? frontend? kernel spawner?) and handles the requests.

**Comment by echarles** [comment]
> it could directly call an S3 API without passing through the Jupyter contents API

True, I would say we need to not take into account that case, and stick to the canonical Jupyter architecture.

> I opened an https://github.com/jupyter-server/jupyter_server/issues/1115 about exposing the contents API to kernels.

Looks like it is closed. I found interesting to connect from the kernel to the contents API.

> For an exploration of the opposite problem - accessing/finding out who (server? client? remote drive?) owns files from the frontend space - see https://github.com/jupyter-server/jupyter_server/issues/1280.

Exciting @krassowski !  You say`accessing/finding out who (server? client? remote drive?) owns files`, I would add `who creates the drives and how the other actors are notified and get access to them?`

**Comment by dhirschfeld** [comment]
I'm wondering if this is relevant for [`jupyter-fs`](https://github.com/jpmorganchase/jupyter-fs) which @timkpaine has been working on recently :thinking:
